src/vs/workbench/contrib/chat/browser/promptFileReference.ts

```diff
@@ -0,0 +1,393 @@
+ *--------------------------------------------------------------------------------------------*/
+
+import { URI } from '../../../../base/common/uri.js';
+import { Emitter } from '../../../../base/common/event.js';
+import { extUri } from '../../../../base/common/resources.js';
+import { Disposable } from '../../../../base/common/lifecycle.js';
+import { Location } from '../../../../editor/common/languages.js';
+import { IConfigurationService } from '../../../../platform/configuration/common/configuration.js';
+import { ChatbotPromptCodec } from '../../../common/codecs/chatbotPromptCodec/chatbotPromptCodec.js';
+import { FileOpenFailed, NotPromptSnippetFile, RecursiveReference } from './promptFileReferenceErrors.js';
+import { FileChangesEvent, FileChangeType, IFileService, IFileStreamContent } from '../../../../platform/files/common/files.js';
+
+/**
+ * Error conditions that may happen during the file reference resolution.
+ */
+export type TErrorCondition = FileOpenFailed | RecursiveReference | NotPromptSnippetFile;
+
+/**
+ * File extension for the prompt snippet files.
+ */
+const PROMP_SNIPPET_FILE_EXTENSION: string = '.prompt.md';
+
+/**
+ * Configuration key for the prompt snippets feature.
+ */
+const PROMPT_SNIPPETS_CONFIG_KEY: string = 'chat.experimental.prompt-snippets';
+
+/**
+ * Represents a file reference in the chatbot prompt, e.g. `#file:./path/to/file.md`.
+ * Contains logic to resolve all nested file references in the target file and all
+ * referenced child files recursively, if any.
+ *
+ * ## Examples
+ *
+ * ```typescript
+ * const fileReference = new PromptFileReference(
+ * 	 URI.file('/path/to/file.md'),
+ * 	 fileService,
+ * );
+ *
+ * // subscribe to updates to the file reference tree
+ * fileReference.onUpdate(() => {
+ * 	 // .. do something with the file reference tree ..
+ * 	 // e.g. get URIs of all resolved file references in the tree
+ * 	 const resolved = fileReference
+ * 		// get all file references as a flat array
+ * 		.flatten()
+ * 		// remove self from the list if only child references are needed
+ * 		.slice(1)
+ * 		// filter out unresolved references
+ * 		.filter(reference => reference.resolveFailed === flase)
+ * 		// convert to URIs only
+ * 		.map(reference => reference.uri);
+ *
+ * 	 console.log(resolved);
+ * });
+ *
+ * // *optional* if need to re-resolve file references when target files change
+ * // note that this does not sets up filesystem listeners for nested file references
+ * fileReference.addFilesystemListeners();
+ *
+ * // start resolving the file reference tree; this can also be `await`ed if needed
+ * // to wait for the resolution on the main file reference to complete (the nested
+ * // references can still be resolving in the background)
+ * fileReference.resolve();
+ *
+ * // don't forget to dispose when no longer needed!
+ * fileReference.dispose();
+ * ```
+ */
+export class PromptFileReference extends Disposable {
```

Comment:
It looks like this is sort of a model object that doesn't depend on anything else in `browser/` so can it go in the `common/` layer?

Reply:
Was thinking the same but wasn't 100% certain, let me try to move it over 🚀

Reply:
Moved.